.cloudbuild/scripts/internal/artifacts/artifacts_test.go

```diff
@@ -0,0 +1,259 @@
+
+	t.Run("Patterns are expanded", func(t *testing.T) {
+		patterns := []string{"alpha", "nested/beta", "*"}
+
+		expected := []string{
+			filepath.Join(workspace, "alpha"),
+			filepath.Join(workspace, "nested", "beta"),
+			filepath.Join(workspace, "*"),
+		}
+
+		actual, err := ValidatePatterns(workspace, patterns)
+		require.NoError(t, err)
+
+		require.ElementsMatch(t, actual, expected)
+	})
+
+	t.Run("Paths are canonicalised", func(t *testing.T) {
+		patterns := []string{
+			"nested/../alpha",
+			"./beta",
+		}
+
+		expected := []string{
+			filepath.Join(workspace, "alpha"),
+			filepath.Join(workspace, "beta"),
+		}
+
+		actual, err := ValidatePatterns(workspace, patterns)
+		require.NoError(t, err)
+
+		require.ElementsMatch(t, actual, expected)
+	})
+
+	t.Run("Paths outside workspace fail", func(t *testing.T) {
+		t.Run("fully-qualified path", func(t *testing.T) {
+			_, err := ValidatePatterns(workspace, []string{t.TempDir()})
+			require.Error(t, err)
+		})
+
+		t.Run("relative path", func(t *testing.T) {
+			target := "../../root/**/*"
+			_, err := ValidatePatterns(workspace, []string{target})
+			require.Error(t, err)
+		})
+	})
+}
+
+func TestFindArtifacts(t *testing.T) {
+	workspace := t.TempDir()
+
+	alpha := touch(t, workspace, "alpha.yaml")
+	beta := touch(t, workspace, "beta.yaml")
+	gamma := touch(t, workspace, "gamma.some-other-extension")
+	delta := touch(t, workspace, "nested", "delta.yaml")
+	epsilon := touch(t, workspace, "nested", "epsilon.yaml")
+	zeta := touch(t, workspace, "nested", "deeply", "zeta")
+	eta := touch(t, workspace, "nested", "very", "deeply", "eta.yaml")
+
+	t.Run("root-dir", func(t *testing.T) {
+		patterns := []string{filepath.Join(workspace, "*")}
+		actual := find(patterns)
+		expected := []string{alpha, beta, gamma}
+		require.ElementsMatch(t, expected, actual)
+	})
+
+	t.Run("prefix", func(t *testing.T) {
+		patterns := []string{
+			filepath.Join(workspace, "nested/*"),
+			filepath.Join(workspace, "nested/*/*"),
+			filepath.Join(workspace, "nested/*/*/*"),
+		}
+		actual := find(patterns)
+		expected := []string{delta, epsilon, zeta, eta}
+		require.ElementsMatch(t, expected, actual)
+	})
+
+	t.Run("suffix", func(t *testing.T) {
+		patterns := []string{
+			filepath.Join(workspace, "*.yaml"),
+			filepath.Join(workspace, "*/*.yaml"),
+			filepath.Join(workspace, "*/*/*.yaml"),
+			filepath.Join(workspace, "*/*/*/*.yaml"),
+		}
+		actual := find(patterns)
+		expected := []string{alpha, beta, delta, epsilon, eta}
+		require.ElementsMatch(t, expected, actual)
+	})
+
+	t.Run("infix", func(t *testing.T) {
+		patterns := []string{filepath.Join(workspace, "*/very/deeply/*")}
+		actual := find(patterns)
+		expected := []string{eta}
+		require.ElementsMatch(t, expected, actual)
+	})
+}
+
+func TestUpload(t *testing.T) {
+	workspace := t.TempDir()
+
+	ctx := context.Background()
+	bucket := new(mockBucket)
+
+	mock := func(content []byte, path ...string) (string, *bytes.Buffer) {
```

Comment:
nit: What is `mock` mocking? Please the function a more descriptive name.

Reply:
done